### ISSUE TITLE
Add support for persistence and running queries on data stored on disk

### ISSUE THREAD

**Issue by cswinter** [cswinter/LocustDB#3]
# Braindump

This is less difficult than it sounds, I think we can leverage RocksDB to most of the heavy lifting.

This work can be split up into a number of incremental and independent steps. I think there will be three big components, RocksDB schema design, interactions with the query engine, and serialization.

## Disk Subsystem
I'm pretty sure we can just make RocksDB work for us, I don't think we will (ever) gain in any notable performance improvements by rolling our own persistent database. LocustDB places very modest demands on the database layer: there is no support for updates, and all reads load fairly large values (megabytes) which should make it fairly easy to design a schema that will give us minimal read amplification.

At least for now, I think we can keep the metadata in memory and then load the data for individual columns on demand as required by running queries. The meta data consist of information of what tables exist, and for each table a list of partition meta data which lists the names all the columns in that partition as well as some summary information on the values contained in each column (right now just min, max).

So the interface for the disk subsystem can be fairly simple, and I think it maps relatively straightforwardly to a RocksDB schema:

```Rust
fn load_column(table_name: &str, partition_id: usize, column_name: &str) -> ColumnData;
// Called after restart to restore all metadata
fn load_metadata() -> Vec<Metadata>;
// persist new data to disk
fn store_column(data: ColumnData, metadata: MetaData);
```

We probably want to have two separate column families for storing metadata and columns to make the initial LoadMetadata() fast, and we can use RocksDB transactions to keep them in sync.

At some point we will want to add a compaction mechanism that merges several partitions into one. This can be implemented by computing the merged partition and then running an atomic transaction that inserts the new partitions and deletes the old ones.

Efficiently supporting very high ingestion rates might require some additional cleverness and tuning of various RocksDB parameters, but I think even a simple design will get us very far.

Once we are storing more data on disk then we have memory, we need the ability to evict data from memory. I think this should be relatively straightforward, but still requires figuring out some questions:
- How to accurately estimate memory usage/free memory.
- With what part of the code lies the responsibility of managing loading/eviction of memory cached data and how exactly does that tie in with the query engine, scheduler and disk subsystem.

## Scheduling
At the moment the query scheduling in LocustDB is very simple. At start-up, we spawn one worker thread for each core. All worker threads wait on a single task queue. When they pick up a new task, they will run it until completion.

Once we have queries that actually read from disk, we will want to have the ability to free up threads to work on different queries while data is loaded from disk (but the original implementation can just block). I'm not super clear on how all of this would work, we can probably leverage the new async/futures stuff. How can we fit it in with the existing query engine and task scheduling? Maybe we just need to completely redesigned that part to work well with Futures.

## Serialization
The initial implementation might just convert everything to a single common data format, and then perform the same processing/compression that is done during ingestion. But that leaves a lot of performance on the table, if we can persist the compressed data to disk in the internal format we will gain large improvements in cold query speed, disk usage, and restart time.

The interface to RocksDB is just bytes, and we should require all of our in memory compression schemes to be converted to/restored from that format with minimal overhead. The serialized column will probably look something like this:

```
Version                       # Allows us to make backwards compatible changes to the format
Compression algorithm         # Tells us how the data can be decoded
One or more data sections     # The actual data
```

We can probably use some serialization library to actually figure out the layout (which serialization library though).

I think this is relative straightforward in theory, but will still be a bit tricky to get right. We will need some kind of framework that ensures that all data is routed to right compression algorithm and makes it easy to add new compression algorithms or modify existing ones all while maintaining backwards compatibility.

**Comment by cswinter** [comment]
I had some neat ideas and can't help myself so I'll have a go at this.

**Comment by cswinter** [comment]
Working prototype for persisting to disk and lazily loading data in https://github.com/cswinter/LocustDB/commit/0cf5df1e715dbdde24f70740c1041f56cd15116b. Needs some more cleanup, but performance is already looking excellent.

**Comment by cswinter** [comment]
#18 adds a working storage backend and figures out a good approach for serialization. Left for future work:
- scheduling
- non-blocking io
- evict in-memory data
- use backwards compatible serialization scheme (e.g. cap'n proto)
- optimizations
